p2p/src/main/kotlin/maru/p2p/MaruPeerManagerService.kt

```diff
@@ -0,0 +1,133 @@
+  p2pConfig: P2P,
+) : PeerHandler {
+  private val log: Logger = LogManager.getLogger(this::class.java)
+
+  private val maxPeers = p2pConfig.maxPeers
+  private val currentlySearching = AtomicBoolean(false)
+  private val connectionInProgress = mutableListOf<Bytes>()
+
+  private var discoveryService: MaruDiscoveryService? = null
+  private lateinit var p2pNetwork: P2PNetwork<Peer>
+  private var stopCalled = false
+
+  fun start(
+    discoveryService: MaruDiscoveryService?,
+    p2pNetwork: P2PNetwork<Peer>,
+  ) {
+    this.discoveryService = discoveryService
+    this.p2pNetwork = p2pNetwork
+    searchForPeersUntilMaxReached()
+  }
+
+  fun stop(): SafeFuture<Unit> {
+    stopCalled = true
+    return SafeFuture.completedFuture<Unit>(null)
+  }
+
+  private fun searchForPeersUntilMaxReached() {
+    if (!stopCalled && currentlySearching.compareAndSet(false, true)) {
+      discoveryService?.let { discoveryService ->
+        discoveryService
+          .searchForPeers()
+          .orTimeout(Duration.ofSeconds(30L))
+          .whenComplete { availablePeers, throwable ->
+            log.debug("Finished searching for peers. Found {} peers.", availablePeers.size)
+            if (!stopCalled) {
+              if (throwable != null) {
+                log.debug("Failed to discover peers: {}\n{}", throwable.message, throwable.stackTraceToString())
+                discoveryService.getKnownPeers().forEach { peer ->
+                  tryToConnectIfNotFull(peer)
+                }
+              } else {
+                availablePeers.forEach { peer ->
+                  tryToConnectIfNotFull(peer)
+                }
+              }
+            }
+          }.whenComplete { _, _ ->
+            currentlySearching.set(false)
+            log.debug("Peer count: {}. Max Peers: {}", p2pNetwork.peerCount, maxPeers)
+            if (!stopCalled && p2pNetwork.peerCount < maxPeers) {
+              CompletableFuture.runAsync { searchForPeersUntilMaxReached() }
+            }
+          }
+      }
+    }
+  }
+
+  override fun onConnect(peer: Peer) {
+    // TODO: here we could check if we want to be connected to that peer
+    if (p2pNetwork.peerCount > maxPeers) {
+      // TODO: We could disconnect another peer here, based on some criteria
+      peer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS)
+    }
+  }
+
+  override fun onDisconnect(peer: Peer) {
+    if (!stopCalled && p2pNetwork.peerCount < maxPeers) {
+      searchForPeersUntilMaxReached()
+    }
+  }
+
+  private fun tryToConnectIfNotFull(peer: MaruDiscoveryPeer) {
+    try {
+      synchronized(connectionInProgress) {
+        if (stopCalled) {
+          return
+        }
+        if (p2pNetwork.peerCount >= maxPeers || connectionInProgress.contains(peer.nodeIdBytes)) {
+          return
+        }
+        connectionInProgress.add(peer.nodeIdBytes!!)
+      }
+
+      log.debug("Peer {} Connecting to peer {}...", discoveryService?.getLocalNodeRecord()?.nodeId, peer.nodeIdBytes)
```

Comment:
If execution got here, it means that `discoveryService` isn't null. So we better make it clear for the reader of the code by using !!

Reply:
yes